lib/spack/llnl/util/filesystem.py

```diff
@@ -764,59 +766,59 @@ def __init__(self, inner_exception, outer_exception):
 
 @contextmanager
 @system_path_filter
-def replace_directory_transaction(directory_name, tmp_root=None):
-    """Moves a directory to a temporary space. If the operations executed
-    within the context manager don't raise an exception, the directory is
-    deleted. If there is an exception, the move is undone.
+def replace_directory_transaction(directory_name):
+    """Temporarily renames a directory in the same parent dir. If the operations
+    executed within the context manager don't raise an exception, the renamed directory
+    is deleted. If there is an exception, the move is undone.
 
     Args:
         directory_name (path): absolute path of the directory name
-        tmp_root (path): absolute path of the parent directory where to create
-            the temporary
 
     Returns:
         temporary directory where ``directory_name`` has been moved
     """
     # Check the input is indeed a directory with absolute path.
     # Raise before anything is done to avoid moving the wrong directory
-    assert os.path.isdir(directory_name), \
-        'Invalid directory: ' + directory_name
-    assert os.path.isabs(directory_name), \
-        '"directory_name" must contain an absolute path: ' + directory_name
-
-    directory_basename = os.path.basename(directory_name)
-
-    if tmp_root is not None:
-        assert os.path.isabs(tmp_root)
+    directory_name = os.path.abspath(directory_name)
+    assert os.path.isdir(directory_name), 'Not a directory: ' + directory_name
+
+    # Try to rename
+    for _ in range(10):
+        suffix = '_' + ''.join(random.choice(string.ascii_lowercase) for _ in range(5))
+        # normalized path does not have trailing dir separators
+        backup_dir = directory_name + suffix
+        try:
+            os.rename(directory_name, backup_dir)
+            last_error = None
+            break
+        except OSError as e:
+            last_error = e
+            continue
 
-    tmp_dir = tempfile.mkdtemp(dir=tmp_root)
-    tty.debug('Temporary directory created [{0}]'.format(tmp_dir))
+    if last_error is not None:
+        raise last_error
```

Comment:
This is basically `mktemp`, which is deprecated b/c it leads to races (yep I know there were more races before). I think it's better to avoid the custom logic and use `mkdtemp` as we were doing before, BUT force it to make a hidden directory alongside the installation sort of like this:

```python
tmp_dir = tempfile.mkdtemp(dir=os.path.dirname(directory_name), prefix='._overwrite_')
```

This avoids races bt/w finding the path name and creating a destination, as the new directory blocks subsequent `mkdtemp`s.

I am actually not 100% sure why we used `dir=tmp_root` in #5384, but I was surprised when I saw that this wasn't choosing a directory guaranteed to be in the same filesystem as the overwritten path.  I think just removing the `tmp_root` parameter and always choosing something next to the target directory will fix that more cleanly.

Reply:
I guess if you like the loop better than we could make it a routine in `llnl.util.filesystem`, but it occurs to me that there is already an issue with using `rename` in a loop like this, so it would need to be improved.

On Linux, if `backup_dir` already exists and is a directory (either because of a prior failed/in-progress `install --overwrite`, or for who knows what reason) you'll end up moving the old installation directory into the existing directory, and it may be removed by some other process, which will prevent you from restoring it if the new install fails.

You could do what you want with [`renameat2`](https://manpages.debian.org/testing/manpages-dev/renameat2.2.en.html), but it's not widely supported -- @becker33 and I tried to get that working for atomic view updates for environments and went with symlinks instead b/c there's no builtin `renameat2` shim.

Reply:
`rename` works fine as long as the target is empty:

```
tmp.zytAjZQF $ mkdir a b
tmp.zytAjZQF $ touch a/file1 b/file2
tmp.zytAjZQF $ python3 -c 'import os; os.rename("a", "b")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
OSError: [Errno 66] Directory not empty: 'a' -> 'b'
tmp.zytAjZQF $ rm b/file2 
tmp.zytAjZQF $ python3 -c 'import os; os.rename("a", "b")'
tmp.zytAjZQF $ ls b
file1
tmp.zytAjZQF $ ls
b```